[benchmark] SR-10855: Added non-ASCII characters to String benchmarking

Armenm opened the pull request:
* Added Cyrillic and Armenian names.
* Added test for pathological case of Eszett (ß -> SS).
* Added variants that work on longer string (tests for `_SmallString` optimization have `.Small` suffix).

Resolves [SR-10855](https://bugs.swift.org/browse/SR-10855) 
Subtask of [SR-8905: Gaps in String benchmarking](https://bugs.swift.org/browse/SR-8905)

Armenm committed:
SR-10855: Added non-ASCII characters

Added Cyrillic and Armenian names with emojis

jrose-apple commented:
@swift-ci Please benchmark

jrose-apple commented:
@swift-ci Please smoke test

Armenm committed:
Decreased the workload of benchmarks

Armenm commented:
Hey @gottesmm I have decreased the workload, should be enough to pass the check. Could you please re-run the benchmark test?

brentdax commented:
@swift-ci Please benchmark

brentdax commented:
@swift-ci Please smoke test

garricn commented:
🥳🇦🇲👏🏽🤓💪🏽💻📲

brentdax commented:
@Armenm Looks like you still need to reduce the workload a little more. Maybe mixed strings with emoji should be measured in separate benchmarks?

palimondo commented on benchmark/single-source/AngryPhonebook.swift:
```diff
@@ -15,11 +15,21 @@
 import TestsUtils
 import Foundation
 
```
Let's extract the tags:
````swift
let t: [BenchmarkCategory] = [.validation, .api, .String]
````

palimondo commented on benchmark/single-source/AngryPhonebook.swift:
```diff
@@ -15,11 +15,21 @@
+    runFunction: run_AngryPhonebook,
+    tags: [.validation, .api, .String],
+    legacyFactor: 7),
+  BenchmarkInfo(
```
Minor nitpick about [naming convention](https://github.com/apple/swift/blob/master/benchmark/Naming.md) is to separate the script name by a period.

Bellow we'll define workloads for different scripts and extract the common benchmarking function so that we can do this:
````swift
  BenchmarkInfo(
    name: "AngryPhonebook.Latin",
    runFunction: { angryPhonebook($0, latin) },
    tags: t, setUpFunction: { blackHole(latin) }),
  BenchmarkInfo(
    name: "AngryPhonebook.Cyrillic",
    runFunction: { angryPhonebook($0, cyrillic) },
    tags: t, setUpFunction: { blackHole(cyrillic) }),
  BenchmarkInfo(
    name: "AngryPhonebook.Armenian",
    runFunction: { angryPhonebook($0, armenian) },
    tags: t, setUpFunction: { blackHole(armenian) })

palimondo commented on benchmark/single-source/AngryPhonebook.swift:
```diff
@@ -39,3 +49,39 @@ public func run_AngryPhonebook(_ N: Int) {
     }
   }
 }
+
```
Set the workload baseline by the slowest benchmark to be under 1000 μs, looking at the last measured performance of `Armenian` (27^2 = 2636 name pairs), I'm guessing we could standardize on workload size of 20 names per script. We should also try to keep the overall size roughly the same. Document this:
````swift
// Workloads for various scripts. Always 20 names for 400 pairings.
// To keep the performance of various scripts roughly comparable, aim for
// a total length of approximately 120 characters.
// E.g.: `latin.joined(separator: "").count == 124`

let latin = Array(words.prefix(20))

let cyrillic = [
  "Александр", "Аркадий", "Аня", "Даниил", "Дмитрий", "Эдуард", "Юрій", "Давид",
  "Анна", "Дмитрий", "Евгений", "Борис", "Владимир", "Артур", "Антон",
  "Антон", "Надія", "Алёна", "Алиса"]

let armenian = [
  "Արմեն", "Աննա", "Հարություն", "Միքայել", "Մարիա", "Դավիթ", "Վարդան", "Նարինե",
  "Հռիփսիմե", "Տիգրան", "Տաթև", "Ադամ", "Ազատ", "Ազնաւուր", "Գրիգոր", "Անի",
  "Լիլիթ", "Հայկ", "Անդրանիկ", "Գառնիկ"]
````
I think `cyrillic` needs a bit tuning, the first 20 names I took are only 108 characters… could you please bring it up to ~120? Armenian at 114 are fine, I think…

milseman replied:
Armenian and Cyrillic are both 2-byte encoded languages. Unless there are salient differences in how capitalization works (I don't know either enough to know), normalization, or use of combining characters, these would be the same workload from a technical perspective.

As @palimondo pointed out, what's also interesting would be the following:

let smallTwoByte = [
  // something up to 15 UTF-8 code units in length, i.e. 7 "characters" in Cyrillic or Armenian ...
]
let largeTwoByte = [
  // something > 16 UTF-8 code units in length, i.e. at least 7 "characters" in Cyrillic or Armenian...  
]

And the same for 3-byte for some latter-BMP language with capitalization (not sure off the top of my head), and without (e.g. Han ideographs), and similarly for 4-byte, and a large 1-byte (i.e. ASCII) workload. This can be done as a separate PR, though.

palimondo commented on benchmark/single-source/AngryPhonebook.swift:
```diff
@@ -39,3 +49,39 @@ public func run_AngryPhonebook(_ N: Int) {
+  "Антон 👍🏻", "Надія 👍", "Алёна", "Алиса", "Елена 🇷🇺", "Елизавета 👍🏾", "👍🏿 Инна",
+  "Жанна 🇨🇦", "Ульяна", "Кристина", "Ксения", "👍🏿👍🏾👍🏽👍🏼👍🏻👍🇺🇦🇷🇺🇨🇦🇺🇸👨‍👩‍👧‍👦"]
+
+@inline(never)
```
* Reuse the common code:
````swift
@inline(never)
public func angryPhonebook(_ N: Int, _ names: [String]) {
  assert(words.count == 20)
  // Permute the names.
  for _ in 1...N {
    for firstname in names {
      for lastname in names {
        blackHole((firstname.uppercased(), lastname.lowercased()))
      }
    }
  }
}
````
The `blackHole` function is there to prevent the optimizer for eliminating the code if it becomes too smart in the future and would realize it is unused… 😜

palimondo reviewed (changes requested):
Welcome to Swift project, Armen! Looks like we need to coordinate two PRs modifying the same benchmark… See PR #25305. You have filed [SR-10855](https://bugs.swift.org/browse/SR-10855) first, so we'll start here.

Since the addition of other [bicameral scripts](https://en.wikipedia.org/wiki/Letter_case#Bicameral_script) is turning a single benchmark into a family of benchmarks, we should take this opportunity to properly design it to allow for their relative comparison, while we're at it. We'll leave the original `AngryPhonebook` benchmark without changes.

cc @mdab121

Armenm commented:
@palimondo Thanks for the great feedback and support. Yes, I will review and make all necessary changes.
Regarding the Benchmark_Driver.BenchmarkDoctor. Could you please share instructions on how to run it on a local machine. It would be easier for me to contribute to other benchmarks. 
cc @gottesmm

Thanks in advance.

palimondo commented:
Sure, `BenchmarkDriver` has 3 subcommands:
````bash
$ ./Benchmark_Driver --help
usage: Benchmark_Driver [-h] COMMAND ...

optional arguments:
  -h, --help  show this help message and exit

Swift benchmark driver commands:
  COMMAND     See COMMAND -h for additional arguments
    run       Run benchmarks and output results to stdout
    check
    compare   Compare benchmark results

Example: ./Benchmark_Driver run -i 5 -f Prefix -f .*Suffix.*
````
You want to run the check to validate that your newly added benchmarks pass our sanity checks: 
````bash
$ ./Benchmark_Driver check --help
usage: Benchmark_Driver check [-h] [-f PATTERN] [-t TESTS] [-o OPT] [-v | -md]
                              [BENCHMARK [BENCHMARK ...]]

positional arguments:
  BENCHMARK             benchmark to run (default: all)

optional arguments:
  -h, --help            show this help message and exit
  -f PATTERN, --filter PATTERN
                        run all tests whose name match regular expression
                        PATTERN, multiple filters are supported
  -t TESTS, --tests TESTS
                        directory containing Benchmark_O{,none,size} (default:
                        DRIVER_DIR)
  -o OPT, --optimization OPT
                        optimization level to use: {O,Onone,Osize}, (default:
                        O)
  -v, --verbose         show more details during benchmark analysis
  -md, --markdown       format report as Markdown table
````

So for the testing of new `AngryPhonebook` variants you'd run them with
````bash
$ ./Benchmark_Driver run -i 3 --output-dir log -f Angry
````
The `--output-dir` parameter triggers (as a side effect) a nicer result formatting (while writing the CSV formatted log into the `log` directory). The `-i 3` will measure 3 independent runs of each benchmark. The `-f Pattern` will run all benchmarks that match a regex.

To run the benchmark validation you would do:
```bash
$ ./Benchmark_Driver check -f Angry
```
This takes a while, because it measures each benchmark for a second, ten times.

palimondo commented:
cc @mdab121, you mind find these tips for local benchmark validation @Armenm requested also useful… 😉

palimondo commented:
@Armenm I've noticed one more thing, that we could take care of here if we want to be super fair when comparing across different scripts… The `String` has an optimization for small strings that can be packed as a value into the object body to avoid reference counting overhead. You can test it in the REPL of your locally built swift in `bin` directory (which can access also internal methods and properties):
````bash
$ ./swift
***  You are running Swift's integrated REPL,  ***
***  intended for compiler and stdlib          ***
***  development and testing purposes only.    ***
***  The full REPL is built as part of LLDB.   ***
***  Type ':help' for assistance.              ***
(swift) "smol"._guts.isSmall
// r0 : Bool = true
(swift) "Long String > 16B"._guts.isSmall
// r1 : Bool = false
````
So to be totally fair, we should probably take the extra step to ensure we'll be comparing apples to apples, because in latin, all names end up being small strings:
````swift
(swift) latin.allSatisfy { $0._guts.isSmall }
// r7 : Bool = true
````
but 
````swift 
(swift) cyrillic.filter { !$0._guts.isSmall }
// r5 : [String] = ["Александр", "Владимир"]
(swift) armenian.filter { !$0._guts.isSmall }
// r7 : [String] = ["Հարություն", "Հռիփսիմե", "Ազնաւուր", "Անդրանիկ"]
````
Could you try to pick the names with lengths so that they would fit into small strings (16 bytes when UTF-8 encoded) and still get the overall length?

@milseman Do you want to have benchmarks per script (this PR adds Cyrillic and Armenian, PR #25305 is poised to add some Extended Latin)? We could probably add Greek to cover all major bicameral scripts… or do you want just one non-ASCII addition?

Armenm commented:
Hey @palimondo This is the first time I am touching this code and had looked only for 4-5 hours during try! Swift San Jose. Forgive me if it takes too much time to understand what's going on here :)
Really appreciate your help.

palimondo commented:
@Armenm Don't worry, you're doing great and I'm overthinking things....

milseman reviewed (commented):
This is looking great, thanks for working on this! Could you incorporate @palimondo 's code review feedback (naming convention, extracting out the common code), and we can merge this?

Armenm commented:
Sure @milseman Will look into it. Thanks!

Armenm committed:
SR-10855: Consistent test data and code refactor

Armenm commented:
Hey @palimondo. I have refactored the code based on your comments. There are 20 names in each language and ~120 characters.
`latin.joined(separator: "").count = 118`
`cyrillic.joined(separator: "").count = 119`
`armenian.joined(separator: "").count = 116`

Armenm commented:
> To run the benchmark validation locally you would do:
> 
> ```shell
> $ ./Benchmark_Driver check -f Angry
> ```
> 
> This takes a while, because it measures each benchmark for a second, ten times. You'll get color coded warnings and errors if we find something suspicious. To see all the nitty gritty while it does its thing do:
> 
> ```shell
> $ ./Benchmark_Driver check -f Angry --verbose
> ```

I was not able to run `Benchmark_Driver` locally. Could you please check the error below and give advice?

```
./Benchmark_Driver check -f Angry --verbose
Traceback (most recent call last):
  File "./Benchmark_Driver", line 774, in <module>
    exit(main())
  File "./Benchmark_Driver", line 770, in main
    return args.func(args)
  File "./Benchmark_Driver", line 576, in run_check
    doctor = BenchmarkDoctor(args)
  File "./Benchmark_Driver", line 343, in __init__
    self.driver = driver or BenchmarkDriver(args)
  File "./Benchmark_Driver", line 59, in __init__
    self.tests = tests or self._get_tests()
  File "./Benchmark_Driver", line 112, in _get_tests
    self._invoke(self._cmd_list_benchmarks).split('\n')[1:-1]
  File "./Benchmark_Driver", line 68, in _invoke
    cmd, stderr=self._subprocess.STDOUT)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 216, in check_output
    process = Popen(stdout=PIPE, *popenargs, **kwargs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

palimondo commented:
> I was not able to run Benchmark_Driver locally. Could you please check the error below and give advice?

It's unable to find the benchmark binary. You need to be running it from the `bin` directory that also contains `Benchmark_O`.

palimondo commented:
@swift-ci please benchmark

palimondo commented on benchmark/single-source/AngryPhonebook.swift:
```diff
@@ -15,26 +15,54 @@
 
-let words = [
+public let AngryPhonebook = [
+  BenchmarkInfo(
```
As I've said before:
> We'll leave the original `AngryPhonebook` benchmark without changes.

The reason is that we use these benchmarks for long term performance tracking between releases.

Please add the original `AngryPhonebook` back as it was. We can add the additional `AngryPhonebook.Latin` for direct comparison with the other additions. So I'd really go with `let latin = Array(words.prefix(20))` for short.

Armenm replied:
Sorry, missed that.
I guess I need to return the `legacyFactor` too. It makes a huge difference.
And maybe not use the `setUpFunction` parameter to. What do you think? 
The previous implementation didn't have that.

Before
```
BenchmarkInfo(
  name: "AngryPhonebook",
  runFunction: run_AngryPhonebook,
  tags: [.validation, .api, .String],
  legacyFactor: 7)
```
After
```
BenchmarkInfo(
    name: "AngryPhonebook",
    runFunction: { angryPhonebook($0, words) },
    tags: t,
    legacyFactor: 7),
```

Is that good?

palimondo replied:
I suspect replacing the old `run_AngryPhonebook` with the shared `angryPhonebook` will be a measurable difference due to the use of `blackHole`… I'll kick the benchmarks and we'll see.

palimondo replied:
No measurable change. I think we are good to go as is.

palimondo reviewed (changes requested):
We should not modify existing benchmarks due to long term performance tracking.

Armenm committed:
Revert changes to existing benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@eeckstein @shahmishal Did something change in how the CI runs benchmarks? The latest report does not include `Onone` and also the **Benchmark Check Report** is missing. [Yesterday's report looked normal](https://github.com/apple/swift/pull/25309#issuecomment-506365588).

palimondo commented:
@eeckstein Any idea why there are so many false changes reported again here? Is it just that we are adding to `AngryPhonebook` module, which comes as a second benchmark and it messes with alignment of many subsequent modules? I find this surprising… did somehow the fix that performed the module alignment for benchmarks stop working?

palimondo reviewed (approved):
<abbr title="Looks good to me">LGTM</abbr>

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@shahmishal @eeckstein There is definitely something funky going on here on the CI machine that's running benchmarks for this PR. Over at the [other PR](https://github.com/apple/swift/pull/25853), the report includes **-Onone**, **HW Overview** and **Benchmark Check Report**.

palimondo commented:
@swift-ci clean benchmark

palimondo commented:
@swift-ci clean benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@shahmishal @eeckstein Any idea why is this benchmark crashing in `-Onone`?

eeckstein commented:
@palimondo No idea. Does this reproduce locally? If yes, it should be not too hard to debug.

palimondo commented:
Let me see...

palimondo commented:
Ah, OK. It is the `assert(name.count == 20)` which is not true for the original benchmark with all `words`. And asserts are NOPs in optimized builds.

palimondo committed:
[benchmark] Restore original run_AngryPhonebook

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo committed:
[benchmark] Large variants of new AngryPhonebook

Add case conversion benchmarks that operate on one large string (precomposed phonebook, comma separated names).

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@milseman since I was fixing the error and had these benchmarks running locally, I've added a `.Large` variant that tests the case conversion on a large string (~6000 characters ±150), since the originals operate on `_SmallString`s. Could you please review?

palimondo commented:
@Armenm, for future work in this area, you might be interested in what @milseman said in @mdab121's PR https://github.com/apple/swift/pull/25305#issuecomment-505967002

----

@milseman Do you have any explanation for the performance of `Latin.Large`?

Per iteration, the small variants are doing 800 calls to ICU on 2-byte variants (Armenian, Cyrillic), vs. 2 call in the `.Large`, processing roughly the same amount of bytes (large has additional spaces and commas), so it makes sense the `.Large` would be faster compared to small.

The `AngryPhonebook.Latin` is doing the case conversion on the optimized small ASCII strings, so this is clearly the best case scenario, but why would the `AngryPhonebook.Latin.Large` be ~3x slower that the other `.Large` variants? Profiling in Instruments I'm seeing a overhead from `swift_bridgeObjectRetain`… is that to be expected or should I file a bug?

milseman commented:
@palimondo The naming here is preventing me from being able to discuss this intelligently. Are we using "Latin" (which is an irrelevant designation for benchmark purposes) to mean that it consists entirely of ASCII? If so, we should use "ASCII" for the name. Similarly, are we using "Large" to mean that it will not fit in a small-string representation? Otherwise, we're just rearranging the deck chairs: more quicker iterations vs fewer slower iterations.

palimondo commented:
Correct, `Latin` is ASCII. Named this way in expectation of adding `ExtendedLatin` in #25305. I’ve assumed using names of Unicode blocks would not be confusing…

The `Large` variant suffix is added to distinguish it from original `AngryPhonebook` which uses small strings only. We could flip it to use the base name for large variant and add `.Small` suffix for those that use small strings if you prefer?

palimondo commented:
@milseman
> Otherwise, we're just rearranging the deck chairs: more quicker iterations vs fewer slower iterations.

I don’t understand your point here. I was saying that the `Latin.Large` variant of this benchmark has uncovered that a normal string that `isASCII` performs about 2.5x slower than comparably sized non-ASCII String, because it apparently does the conversion in ObjC and not via the faster ICU. It look like there could be easy performance win just by using ICU in that case. Does that make sense?

palimondo commented:
@milseman
> Are we using "Latin" (which is an irrelevant designation for benchmark purposes) to mean that it consists entirely of ASCII?

It seems we have different assumption about the implementation of optimizations for case conversion here. If I understand your perspective from an earlier discussion above it makes sense to distinguish only between cases based on encoding length:
* 1-byte, ASCII
* 2-byte, Extended Latin, Armenian, Cyrillic — all the same?

As I’ve explained in https://github.com/apple/swift/pull/25305#issuecomment-501253740 it seem like could be possible to create optimizations for case conversions for different Unicode code blocks. Is that a wrong way to think about this?

palimondo committed:
[benchmark] Add AngryPhonebook.Strasse variant

 Pathological case, uppercase: ß -> SS

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@milseman I've added a benchmark for the pathological case of uppercased ß -> SS:
````swift
"Straße".uppercased() == "STRASSE"
````

You can see that even that one (`Strasse.Large`) is almost 4 times faster variant than the `Latin.Large`.

milseman commented:
> Correct, Latin is ASCII. Named this way in expectation of adding ExtendedLatin in #25305. I’ve assumed using names of Unicode blocks would not be confusing…

In general, Unicode code blocks is an irrelevant categorization for benchmarking, and “Latin” isn’t even the formal name of the code block (nor would the formal name be as recognizable). ASCII vs non-ASCII is a very relevant categorization, which just so happens to align with a Unicode code block because Unicode explicitly chose to be a superset of ASCII.

For example, a workload that consists entirely of Han Ideographs would be interesting, not because they all come from a given block, but because of the attributes they have in comparison to other workloads, such as case invariance (and to a lesser extent, encoding width for the BMP ones).

Similarly, improvements and additional fast-paths would apply based on these kinds of attributes, such as over case-invariant ranges of scalar values. The boundaries of such ranges are not strictly aligned with code block boundaries.

> The Large variant suffix is added to distinguish it from original AngryPhonebook which uses small strings only. We could flip it to use the base name for large variant and add .Small suffix for those that use small strings if you prefer?

I was confused by your workload in https://github.com/apple/swift/pull/25309/files#r291816905, which had “small” workloads that don’t fit in the small-string representation (15 bytes of UTF-8). If the PR is consistent in small vs large aligning with small vs large string representations, then this is fine. Otherwise I think that “long” vs “short” is better.


> I don’t understand your point here. I was saying that the Latin.Large variant of this benchmark has uncovered that a normal string that isASCII performs about 2.5x slower than comparably sized non-ASCII String, because it apparently does the conversion in ObjC per character(?) and not via the faster ICU. It look like there could be easy performance win just by using ICU in that case. Does that make sense?

This seems unlikely, and perhaps we need to dig into the details if this is happening. Interfacing with ICU requires that we:

1. Allocate a buffer (e.g. `Array`) of `UInt16`s and transcode our contents into it.
2. Allocate another buffer for ICU to write the result to.
3. Call ICU
4. See if our prior buffer was big enough (and redo with a larger one if not)
5. Create a new String (heap allocation if large) while transcoding the UTF-16 back to UTF-8.

In contrast, if we know a String is all-ASCII, we do the following:

1. Allocate a String storage of sufficient capacity
2. Copy the lowercased bytes in

Point #2 could probably be improved with wider/bulk operations.

> 2-byte, Extended Latin, Armenian, Cyrillic — all the same?

If Extended Latin, Armenian, and Cyrillic have similar/identical case-conversion properties (e.g. invariance), and ICU’s implementation is otherwise not distinguishing between them in its own fast paths, then yes.


> As I’ve explained in #25305 (comment) it seem like it could be possible to create optimizations for case conversions specialized for different Unicode code blocks. Is that a wrong way to think about this?

Again, code block boundaries is irrelevant. You’re likely going to optimize for ranges of code points that may span or lie within code block boundaries.


> Edit: Looking at Rust's implementation, I see that the case conversions are implemented using a BoolTrie tables. Is this an approach you are planning use?

A BoolTrie or something similar would be very useful for representing Unicode Scalar Properties in the stdlib without asking ICU on a scalar-by-scalar basis. That includes case conversion properties, though we might want to combine it with some kind of sparse interval set for case invariance.

> Am I misunderstanding what is going on here?

That profile is demonstrating something highly suspect about the implementation.

* ` _StringGuts.append(_:)` this might be performing highly-redundant introspection on every byte.

This should switch to https://forums.swift.org/t/pitch-add-a-string-initializer-with-access-to-uninitialized-storage/22611 when it happens. @Catfish-Man, here is another example of something that would benefit from this. We should propose it immediately after 5.1 wraps up.

* `swift_bridgeObjectRetain` should early exit in the runtime, since small strings are immortal.
* `_swift_retain_(swift::HeapObject*)` should not be happening at all inside here.

Something is going terribly wrong in the implementation (or benchmark construction, but I suspect the implementation) that’s sending us down a *very* slow path here.

* `specialized String.uppercased()` Why is this “specialized”?

The difference between these two traces for `uppercased()` is ~1200, which should be easily recoverable or surpassed. Why is the difference between `largeAngryPhonebook` and `uppercased()` ~2000 in one run and ~600 in the other? That seems suspect as well to me.

palimondo committed:
[benchmark] Cleaned up new AngryPhonebook tests

Clarified the naming of tests that exercise `_SmallString`s with `.Small` suffix and used plane name for regular String workloads.
Reordered declarations. Documented workload sizes.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@milseman 
> The difference between these two traces for `uppercased()` is ~1200, which should be easily recoverable or surpassed. Why is the difference between `largeAngryPhonebook` and `uppercased()` ~2000 in one run and ~600 in the other? That seems suspect as well to me.

In addition to `uppercased()` the `largeAngryPhonebook` also contains call to `lowercased()` which takes 1966 in the suspicious ASCII case. In the case of `Strasse`, `uppercased()`takes 827 and `lowercased()` takes 582, which makes sense because the  upper case conversion needs to grow the string.

I have filed a bug for this performance anomaly: [SR-11069](https://bugs.swift.org/browse/SR-11069).

---- 

I have renamed the tests that exercise the `_SmallString` optimization to have `.Small` suffix. The test for regular `String` have just the plain name denoting the Code Block. To ease the relative comparison, the workload sizes are also documented in the source code:

  SMALL  | UTF-8 | UTF-16 |    REGULAR   |  UTF-8  | UTF-16
--------:|------:|-------:|-------------:|--------:|-------:
   ascii | 124 B |  248 B |    longASCII |  6158 B | 12316 B
 strasse | 140 B |  240 B |  longStrasse |  6798 B | 11996 B
armenian | 232 B |  232 B | longArmenian | 10478 B | 11676 B
cyrillic | 238 B |  238 B | longCyrillic | 10718 B | 11916 B

palimondo commented:
@Armenm, I'll merge this as soon as all the tests pass. I hope you were not discouraged by the meandering progress… Thank you for your contribution!

milseman commented:
@palimondo it seems removing all that senseless overhead gives us around a 30x speedup for the large case, and nice wins even for the smaller one.

https://github.com/apple/swift/pull/26007#issuecomment-509420261

Armenm commented:
> @Armenm, I'll merge this as soon as all the tests pass. I hope you were not discouraged by the meandering progress… Thank you for your contribution!

Not discouraged at all @palimondo. I would say even the opposite. It was really cool to dive deep into the details. And I am happy to see that my humble help brought to some other implementation and optimization on `String`.
Thanks everyone for the support and making Swift a better language. I will definitely contribute in the future and encourage everyone to do it too.
